### PR TITLE
Handle all FileTypes in OutputFileMap

### PR DESCRIPTION
**Pull request by sjavora**
This replaces the specific properties in `Entry` with a dictionary keyed by `FileType` and adds a custom implementation of `Codable`.

Originally, I wanted to just use `FileType`'s `description` property, but some of the cases just return `rawValue`, which is not compatible in all cases ("o" vs "object", in particular). Instead I added the `name` property with values from [FileTypes.def](https://github.com/apple/swift/blob/master/include/swift/Basic/FileTypes.def). While there, I noticed that `FileType` doesn't currently have a case for `swiftsourceinfo`, so I added that as well.

### CODE REVIEW

**Comment by sjavora** [comment]
Hey @DougGregor - could someone please take a look at this? I'd request a review, but the button doesn't seem to be there...

File: Sources/SwiftDriver/Utilities/FileType.swift [hunk 1]
```diff
@@ -163,8 +166,81 @@ extension FileType {
...
+         .swiftDeps, .moduleTrace, .tbd, .optimizationRecord, .swiftInterface, .swiftSourceInfoFile:
       return false
     }
   }
 }
+
+extension FileType {
+
+  init?(name: String) {
+    guard let type = FileType.allCases.first(where: { $0.name == name }) else { return nil }
```
Comment (harlanhaskins) [inline Sources/SwiftDriver/Utilities/FileType.swift]: Is there a way we can make this not O(n)?

**Comment by harlanhaskins** [comment]
The approach looks good here, I've got one small nitpick though.

File: Sources/SwiftDriver/Utilities/FileType.swift (hunk 1 above)
Comment (sjavora) [reply Sources/SwiftDriver/Utilities/FileType.swift]: We could have something like `static let typesByName = Dictionary(uniqueKeysWithValues: FileType.allCases.map { ($0.name, $0) })` which would only enumerate once and then use the dictionary in the initializer. Would that be OK?

Comment (harlanhaskins) [reply Sources/SwiftDriver/Utilities/FileType.swift]: Sounds good.

Comment (sjavora) [reply Sources/SwiftDriver/Utilities/FileType.swift]: Ok, pushed.

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
Thanks, @sjavora!

**Pull request merged**